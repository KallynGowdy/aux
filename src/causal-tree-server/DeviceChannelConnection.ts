import { RealtimeChannelInfo } from '@casual-simulation/causal-trees';
import { DeviceConnection } from './DeviceConnection';

/**
 * Defines an interface that represents a connection to a channel.
 */
export interface DeviceChannelConnection {
    info: RealtimeChannelInfo;
}
