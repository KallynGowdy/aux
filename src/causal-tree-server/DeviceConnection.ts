import { DeviceChannelConnection } from './DeviceChannelConnection';
/**
 * Defines an interface that represents a connection to a device.
 * Calling unsubscribe should
 */
export interface DeviceConnection<TExtra> {
    id: string;
    extra: TExtra;
}
