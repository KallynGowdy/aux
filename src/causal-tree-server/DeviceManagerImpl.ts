import { DeviceManager, ConnectedToChannelListener } from './DeviceManager';
import { DeviceConnection } from './DeviceConnection';
import { DeviceChannelConnection } from './DeviceChannelConnection';
import { RealtimeChannelInfo } from '@casual-simulation/causal-trees';
import { findIndex } from 'lodash';
import { SubscriptionLike, Subscription } from 'rxjs';
import { Device } from 'useragent';

export class DeviceManagerImpl implements DeviceManager {
    activeChannels: RealtimeChannelInfo[];

    private _connectedDevices: Map<
        DeviceConnection<any>,
        ActiveDeviceChannelConnection[]
    >;
    private _activeChannels: Map<string, DeviceConnection<any>[]>;
    private _listeners: ConnectedToChannelListener[];

    constructor() {
        this._connectedDevices = new Map();
        this._activeChannels = new Map();
        this._listeners = [];
    }

    get connectedDevices(): DeviceConnection<any>[] {
        return [...this._connectedDevices.keys()];
    }

    getConnectedChannels(
        device: DeviceConnection<any>
    ): DeviceChannelConnection[] {
        let channels = this._connectedDevices.get(device) || [];
        return channels.slice();
    }

    getConnectedDevices(info: RealtimeChannelInfo): DeviceConnection<any>[] {
        let devices = this._activeChannels.get(info.id) || [];
        return devices.slice();
    }

    whenConnectedToChannel(
        listener: ConnectedToChannelListener
    ): SubscriptionLike {
        this._listeners.push(listener);
        return new Subscription(() => {
            const index = this._listeners.indexOf(listener);
            if (index >= 0) {
                this._listeners.splice(index, 1);
            }
        });
    }

    async connectDevice<TExtra>(
        deviceId: string,
        extra?: TExtra
    ): Promise<DeviceConnection<TExtra>> {
        const device = {
            id: deviceId,
            extra: extra,
        };

        this._connectedDevices.set(device, []);

        return device;
    }

    async disconnectDevice<TExtra>(
        device: DeviceConnection<TExtra>
    ): Promise<void> {
        const channels = this.getConnectedChannels(device);

        for (let channel of channels) {
            await this.leaveChannel(device, channel.info);
        }

        this._connectedDevices.delete(device);
    }

    async joinChannel<TExtra>(
        device: DeviceConnection<TExtra>,
        info: RealtimeChannelInfo
    ): Promise<DeviceChannelConnection> {
        let subs: SubscriptionLike[] = [];
        const channel: ActiveDeviceChannelConnection = {
            info: info,
            subs: subs,
        };

        if (this._connectedDevices.has(device)) {
            const list = this._connectedDevices.get(device);
            list.push(channel);
        }
        let list = this._activeChannels.get(info.id);
        if (!list) {
            list = [];
            this._activeChannels.set(info.id, list);
        }
        list.push(device);

        for (let listener of this._listeners) {
            subs.push(...listener(device, channel));
        }

        return channel;
    }

    async leaveChannel<TExtra>(
        device: DeviceConnection<TExtra>,
        info: RealtimeChannelInfo
    ): Promise<void> {
        if (this._connectedDevices.has(device)) {
            const list = this._connectedDevices.get(device);
            const index = findIndex(list, c => c.info.id === info.id);
            if (index >= 0) {
                const channel = list[index];
                channel.subs.forEach(s => s.unsubscribe());
                list.splice(index, 1);
            }
        }
        let list = this._activeChannels.get(info.id);
        if (list) {
            const index = findIndex(list, d => d === device);
            if (index >= 0) {
                list.splice(index, 1);
            }
        }
    }
}

interface ActiveDeviceChannelConnection extends DeviceChannelConnection {
    subs: SubscriptionLike[];
}
