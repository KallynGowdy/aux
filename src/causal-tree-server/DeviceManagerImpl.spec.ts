import { DeviceManagerImpl } from './DeviceManagerImpl';
import { DeviceConnection } from './DeviceConnection';
import { DeviceChannelConnection } from './DeviceChannelConnection';
import { fake } from 'sinon';
import { Subscription } from 'rxjs';

describe('CausalTreeServerImpl', () => {
    let server: DeviceManagerImpl;

    beforeEach(() => {
        server = new DeviceManagerImpl();
    });

    describe('connectDevice()', () => {
        it('should return a connected device', async () => {
            const device = await server.connectDevice('deviceId');

            expect(device).toMatchSnapshot();
        });

        it('should include the given extras', async () => {
            const device = await server.connectDevice('deviceId', {
                extra: true,
            });

            expect(device).toMatchSnapshot();
        });

        it('should add the device to the connected devices list', async () => {
            const device = await server.connectDevice('deviceId', {
                extra: true,
            });

            expect(server.connectedDevices).toEqual([device]);
        });
    });

    describe('disconnectDevice()', () => {
        let device: DeviceConnection<any>;
        beforeEach(async () => {
            device = await server.connectDevice('deviceId', {
                extra: true,
            });
        });

        it('should remove the given device from the list of connected devices', async () => {
            await server.disconnectDevice(device);

            expect(server.connectedDevices).toEqual([]);
        });

        it('should remove the given device from the channels it is connected to', async () => {
            const channel1 = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            const channel2 = await server.joinChannel(device, {
                id: 'test2',
                type: 'number2',
            });

            await server.disconnectDevice(device);

            const channels = server.getConnectedChannels(device);
            const devices1 = server.getConnectedDevices(channel1.info);
            const devices2 = server.getConnectedDevices(channel2.info);

            expect(channels).toEqual([]);
            expect(devices1).toEqual([]);
            expect(devices2).toEqual([]);
        });
    });

    describe('joinChannel()', () => {
        let device: DeviceConnection<any>;
        beforeEach(async () => {
            device = await server.connectDevice('deviceId', {
                extra: true,
            });
        });

        it('should return a channel connection', async () => {
            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            expect(channel).toMatchSnapshot();
        });

        it('should add the channel to the list for the device', async () => {
            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            const channels = server.getConnectedChannels(device);
            expect(channels).toEqual([channel]);
        });

        it('should add the device to the list for the channel', async () => {
            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            const devices = server.getConnectedDevices(channel.info);
            expect(devices).toEqual([device]);
        });

        it('should allow joining channels with the same ID but different types', async () => {
            const channel1 = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });
            const channel2 = await server.joinChannel(device, {
                id: 'test',
                type: 'different',
            });

            const devices1 = server.getConnectedDevices(channel1.info);
            expect(devices1).toEqual([device]);

            const devices2 = server.getConnectedDevices(channel2.info);
            expect(devices2).toEqual([device]);

            const channels = server.getConnectedChannels(device);
            expect(channels).toEqual([channel1, channel2]);
        });
    });

    describe('leaveChannel()', () => {
        let device: DeviceConnection<any>;
        let channel: DeviceChannelConnection;

        beforeEach(async () => {
            device = await server.connectDevice('deviceId', {
                extra: true,
            });

            channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });
        });

        it('should remove the device from the channel list', async () => {
            await server.leaveChannel(device, channel.info);

            const channels = server.getConnectedChannels(device);
            const devices = server.getConnectedDevices(channel.info);

            expect(server.connectedDevices).toEqual([device]);
            expect(channels).toEqual([]);
            expect(devices).toEqual([]);
        });
    });

    describe('whenConnectedToChannel()', () => {
        it('should call the given function when a device connects', async () => {
            const listener = fake.returns([]);
            server.whenConnectedToChannel(listener);

            const device = await server.connectDevice('deviceId', {
                test: true,
            });

            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            expect(listener.called).toBe(true);
            expect(listener.args).toEqual([[device, channel]]);
        });

        it('should dispose the returned subscriptions when the device leaves the channel', async () => {
            const subs = [new Subscription(), new Subscription()];
            const listener = fake.returns(subs);
            server.whenConnectedToChannel(listener);

            const device = await server.connectDevice('deviceId', {
                test: true,
            });

            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            await server.leaveChannel(device, channel.info);

            expect(subs.map(s => s.closed)).toEqual([true, true]);
        });

        it('should return a subscription that unregisters the listener', async () => {
            const listener = fake.returns([]);
            const sub = server.whenConnectedToChannel(listener);

            const device = await server.connectDevice('deviceId', {
                test: true,
            });

            sub.unsubscribe();

            const channel = await server.joinChannel(device, {
                id: 'test',
                type: 'number',
            });

            expect(listener.called).toBe(false);
        });
    });
});
