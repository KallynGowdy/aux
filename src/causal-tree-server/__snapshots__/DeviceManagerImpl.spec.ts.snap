// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CausalTreeServerImpl connectDevice() should include the given extras 1`] = `
Object {
  "extra": Object {
    "extra": true,
  },
  "id": "deviceId",
}
`;

exports[`CausalTreeServerImpl connectDevice() should return a connected device 1`] = `
Object {
  "extra": undefined,
  "id": "deviceId",
}
`;

exports[`CausalTreeServerImpl joinChannel() should return a channel connection 1`] = `
Object {
  "info": Object {
    "id": "test",
    "type": "number",
  },
  "subs": Array [],
}
`;
